<a href="https://colab.research.google.com/github/TrollRider-Kristian/Springboard-AI-Mini-Projects/blob/main/Student_MLE_MiniProject_Deep_Learning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Mini Project: Deep Learning with Keras

In this mini-project we'll be building a deep learning classifier using Keras to predict income from the popular [Adult Income dataset](http://www.cs.toronto.edu/~delve/data/adult/adultDetail.html).

Predicting income from demographic and socio-economic information is an important task with real-world applications, such as financial planning, market research, and social policy analysis. The Adult dataset, sometimes referred to as the "Census Income" dataset, contains a vast amount of anonymized data on individuals, including features such as age, education, marital status, occupation, and more. Our objective is to leverage this data to train a deep learning model that can effectively predict whether an individual's income exceeds $50,000 annually or not.

Throughout this Colab, we will walk you through the entire process of building a deep learning classifier using Keras, a high-level neural network API that runs on top of TensorFlow. Keras is known for its user-friendly and intuitive interface, making it an excellent choice for both beginners and experienced deep learning practitioners.

Here's a brief outline of what we will cover in this mini-project:

1. **Data Preprocessing:** We will start by loading and exploring the Adult dataset.

2. **Building the Deep Learning Model:** We will construct a neural network using Keras, where we'll dive into understanding the key components of a neural network, including layers, activation functions, and optimization algorithms.

3. **Model Training:** With our model architecture in place, we will split the data into training and validation sets and train the neural network on the training data. We will monitor the training process to prevent overfitting and enhance generalization.

4. **Model Evaluation:** After training, we'll assess the performance of our model on the test dataset.

By the end of this tutorial, you will not only have a functional deep learning classifier for income prediction but also gain valuable insights into how to leverage the power of neural networks for solving real-world classification tasks.


In [42]:
!pip install scikeras

In [44]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder, LabelEncoder
from sklearn.preprocessing import MinMaxScaler
from sklearn.compose import ColumnTransformer
from sklearn.model_selection import train_test_split
from sklearn.metrics import roc_auc_score
from sklearn.metrics import RocCurveDisplay
from keras.models import Sequential
from keras.layers import Dense
from scikeras.wrappers import KerasClassifier
from sklearn.pipeline import Pipeline

# For visualization in the EDA section of this mini-project
import matplotlib
import seaborn

You can download the Adult data from the link [here](https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data).

Here are your tasks:

  1. Load the Adult data into a Pandas Dataframe.
  2. Ensure the dataset has properly named columns. If the columns are not read in, assign them by referencing the dataset documentation.
  3. Display the first five rows of the dataset.

In [45]:
DATA_PATH = 'https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data'

# Download the dataset and load it into a pandas DataFrame
# KRISTIAN_NOTE - The dataset entries for each row are separated
# by commas.  However, the column names are not included in the file,
# but Pandas, by default, treats the first row of data as the column
# names.  Pass header = None to circumvent this, per documentation:
# https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html
adult_df = pd.read_csv (DATA_PATH, header = None)

In [46]:
# KRISTIAN_NOTE - Unfortunately, there's no way to extract the column names from
# the dataset itself.  These names are explicitly listed in the Stanford paper
# that has introduced and discussed this dataset at length:
# https://cseweb.ucsd.edu/classes/sp15/cse190-c/reports/sp15/048.pdf
# 'fnlwgt' stands for 'final weight' or "the number of people the census in the entry represents"
# I named the target column "income-level" because it's categorical (over or under $50K annually)
# Our only option is to add hardcoded column names to this dataframe per
# the ones listed in the dataset. StackOverflow article showing how to do this:
# https://stackoverflow.com/questions/72863835/how-to-create-column-names-in-pandas-dataframe
adult_feature_names = [
    "age",
    "workclass",
    "fnlwgt",
    "education",
    "education-num",
    "marital-status",
    "occupation",
    "relationship",
    "race",
    "sex",
    "capital-gain",
    "capital-loss",
    "hours-per-week",
    "native-country",
    "income-level",
]
adult_df.columns = adult_feature_names

In [47]:
# Display the first few rows of the DataFrame
print (adult_df.head())

   age          workclass  fnlwgt   education  education-num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   

        marital-status          occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital-gain  capital-loss  hours-per-week  native-country income-level  
0          2174             0              40   United-States  

If you're not already familiar with the Adult dataset, it's important to do some exploratory data analysis.

Here are your tasks:

  1. Do exploratory data analysis to give you some better intuition for the dataset. This is a bit open-ended. How many rows/columns are there? How are NULL values represented? What's the percentage of positive cases in the dataset?

  2. Drop all rows with NULL values.

  3. Use Scikit-Learn's [LabelEncoder](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.LabelEncoder.html) to convert the `income` column with a data type string to a binary variable.

In [48]:
# Do some exploratory analysis. How many rows/columns are there? How are NULL
# values represented? What's the percentrage of positive cases in the dataset?
print (f"There are: {adult_df.shape [0]} rows and {adult_df.shape [1]} columns in the Adult dataset.")
adult_df["income-level"].value_counts()

There are: 32561 rows and 15 columns in the Adult dataset.


income-level
<=50K    24720
>50K      7841
Name: count, dtype: int64

24,720 adults, or approximately 76% of the adults, make $50K annually or less. while 7841 make more than this annually.

Interestingly, this data does not handle null values as None or NULL or NaN, but rather as '?', which the isnull() function will not pick up below.

In [49]:
adult_df.isnull().sum()

age               0
workclass         0
fnlwgt            0
education         0
education-num     0
marital-status    0
occupation        0
relationship      0
race              0
sex               0
capital-gain      0
capital-loss      0
hours-per-week    0
native-country    0
income-level      0
dtype: int64

Let's pull up the column types first.

In [50]:
adult_df.dtypes

age                int64
workclass         object
fnlwgt             int64
education         object
education-num      int64
marital-status    object
occupation        object
relationship      object
race              object
sex               object
capital-gain       int64
capital-loss       int64
hours-per-week     int64
native-country    object
income-level      object
dtype: object

Now, we get the value_counts for each of the non-integer columns.
The int64 columns probably don't have null values because NaN is a null value that the isnull() function above would have picked up.  See the docs: https://pandas.pydata.org/docs/reference/api/pandas.isnull.html

In [51]:
# KRISTIAN_NOTE - Fortunately, all categorical columns except education-num for
# this dataset have a datatype of object.  Therefore, we don't need to
# explicitly state each and every categorical column in a hardcoded array.
for col in adult_df.select_dtypes (include = 'object').columns:
  print (adult_df[col].value_counts())

workclass
Private             22696
Self-emp-not-inc     2541
Local-gov            2093
?                    1836
State-gov            1298
Self-emp-inc         1116
Federal-gov           960
Without-pay            14
Never-worked            7
Name: count, dtype: int64
education
HS-grad         10501
Some-college     7291
Bachelors        5355
Masters          1723
Assoc-voc        1382
11th             1175
Assoc-acdm       1067
10th              933
7th-8th           646
Prof-school       576
9th               514
12th              433
Doctorate         413
5th-6th           333
1st-4th           168
Preschool          51
Name: count, dtype: int64
marital-status
Married-civ-spouse       14976
Never-married            10683
Divorced                  4443
Separated                 1025
Widowed                    993
Married-spouse-absent      418
Married-AF-spouse           23
Name: count, dtype: int64
occupation
Prof-specialty       4140
Craft-repair         4099
Exec-managerial      

The only indeterminate value in this dataset is '?'.  Had there been other indeterminate values, like 'N/A' or 'Preferred not to disclose', value_counts() would have picked those up too.

We need to replace every instance of '?' with None and then rerun the isnull() function from above, but we hit a snag in that replacing non-null values with None will not actually change the value to None.  In our case, it just doesn't replace the '?' at all.  See this StackOverflow article: https://stackoverflow.com/questions/17097236/replace-invalid-values-with-none-in-pandas-dataframe

In [52]:
# Leaving in this commented code as a lesson on Pandas replace.
# adult_df.replace (to_replace = '?', value = np.nan, inplace = True)
# print (adult_df.isnull().sum())
# print (adult_df['workclass'].value_counts())

Fortunately the aforementnioned StackOverflow article mentioned another solution.  The read_csv() method can be assigned values to treated as null, so we'll pass '?' and give this another go.

In [53]:
# Reload the dataset, but pass the na_values parameter to force it to treat '?' as NULL.
# Don't forget to check for leading spaces: https://stackoverflow.com/questions/56825958/why-does-not-work-properly-na-values-in-pandas
adult_df = pd.read_csv (DATA_PATH, header = None, na_values = ['?', ' ?'])
adult_df.columns = adult_feature_names
adult_df.isnull().sum()

age                  0
workclass         1836
fnlwgt               0
education            0
education-num        0
marital-status       0
occupation        1843
relationship         0
race                 0
sex                  0
capital-gain         0
capital-loss         0
hours-per-week       0
native-country     583
income-level         0
dtype: int64

In [54]:
# Find all NULL values and drop them
# Now that we've reloaded the dataset with the '?' value being treated as NULL,
# we can remove those rows in the standard way.
adult_df.dropna (how = 'any', inplace = True)

In [55]:
print (f"There are: {adult_df.shape [0]} rows and {adult_df.shape [1]} columns in the Adult dataset.")
adult_df.isnull().sum()

There are: 30162 rows and 15 columns in the Adult dataset.


age               0
workclass         0
fnlwgt            0
education         0
education-num     0
marital-status    0
occupation        0
relationship      0
race              0
sex               0
capital-gain      0
capital-loss      0
hours-per-week    0
native-country    0
income-level      0
dtype: int64

In [56]:
adult_df["income-level"].value_counts()

income-level
<=50K    22654
>50K      7508
Name: count, dtype: int64

Note that there were 32,561 rows when we loaded the Adult dataset the first time.  There are now 30,162 rows total, and the isnull() method no longer catches any rows including a '?'.  Of these rows, 22,654 people, or roughly 75%, make less than or equal to $50k.

Before we split the training and test data, more EDA is required to understand the distribution of categorical variables.  This allows us to determine how feasible it is to stratify our train-test split so that every member of our test set belongs to categories already taken into account in the training set.

In [58]:
# Use Scikit-Learn's LabelEncoder to convert the income column with a data type
# string to a binary variable.
le_income = LabelEncoder()
adult_df["income-over-50k"] = le_income.fit_transform (adult_df["income-level"])
adult_df_encoded_target = adult_df.drop (labels = "income-level", axis = 1)
print (adult_df_encoded_target.head(10))

   age          workclass  fnlwgt   education  education-num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   
5   37            Private  284582     Masters             14   
6   49            Private  160187         9th              5   
7   52   Self-emp-not-inc  209642     HS-grad              9   
8   31            Private   45781     Masters             14   
9   42            Private  159449   Bachelors             13   

           marital-status          occupation    relationship    race  \
0           Never-married        Adm-clerical   Not-in-family   White   
1      Married-civ-spouse     Exec-managerial         Husband   White   
2                Divorced   Handlers-cleaners   Not-in-family   White   
3  

In the previous mini-project (please see Student_MLE_MiniProject_ML.ipynb), I dealt with a dataset of taxi records including id's for locations to pick up and drop off passengers.  When splitting that data, there were three taxi rides in the test set featuring pickup location id's not present in the training set.  Because there were so many distinct location id's, it was impossible to stratify the data.  However, there were so few of these unaccounted offenders that removing them from the test set left almost zero impact on the model.

But I do not find this satisfactory.  This test set has only a fraction of the rows as the Taxicab dataset from the previous project.  I'd rather not deal with this problem again.  After having discussed this issue with my mentor, I now know it is common practice to apply one-hot encoding to the dataset before the train-test split step.

However, the ColumnTransformer portion of this project will need to be integrated into the same Pipeline as the Keras model I will be building later in the project, and that Pipeline only takes "transformer" type operations, and scikit-learn's train-test-split operation is NOT a "transformer" type operation.

To resolve this "chicken and the egg" situation, I have decided to do the following:

1. I will use the get_dummies function from Pandas first because it returns a DataFrame that's easy to split.

2. Then, I will train-test-split the dataset.

3. Then, I will measure the effectiveness of a baseline model.

4. Then, I will create the ColumnTransformer, but the OneHotEncoding step will not take any columns into account.

In [102]:
# KRISTIAN_NOTE - Fortunately, all categorical columns except education-num for
# this dataset have a datatype of object.  Therefore, we don't need to
# explicitly state each and every categorical column in a hardcoded array.
# We need only remove 'income-level' since that's our target before label-encoding.
categorical_feature_cols = list(adult_df.select_dtypes (include = 'object').columns[0:-1])
print (categorical_feature_cols)

# Again, remove 'income-over-50k' since that's our target after label-encoding.
quanitative_feature_columns = list (adult_df.select_dtypes (include = 'int64').columns[0:-1])
print (quanitative_feature_columns)

['workclass', 'education', 'marital-status', 'occupation', 'relationship', 'race', 'sex', 'native-country']
['age', 'fnlwgt', 'education-num', 'capital-gain', 'capital-loss', 'hours-per-week']


1. Use get_dummies to one-hot encode the data.
Docs: https://pandas.pydata.org/docs/reference/api/pandas.get_dummies.html

In [103]:
# List out the columns explicitly.
# StackOverflow article for this: https://stackoverflow.com/questions/36285155/pandas-get-dummies
adult_df_converted_cols = pd.get_dummies (adult_df_encoded_target, columns = categorical_feature_cols)
print (adult_df_converted_cols.head())

   age  fnlwgt  education-num  capital-gain  capital-loss  hours-per-week  \
0   39   77516             13          2174             0              40   
1   50   83311             13             0             0              13   
2   38  215646              9             0             0              40   
3   53  234721              7             0             0              40   
4   28  338409             13             0             0              40   

   income-over-50k  workclass_ Federal-gov  workclass_ Local-gov  \
0                0                   False                 False   
1                0                   False                 False   
2                0                   False                 False   
3                0                   False                 False   
4                0                   False                 False   

   workclass_ Private  ...  native-country_ Portugal  \
0               False  ...                     False   
1               

Now, the DataFrame has 105 columns.

2. Split the data into training and test sets. Remember not to include the label you're trying to predict, `income`, as a column in your training data.

In [104]:
# Split dataset into training and test sets
features_train, features_test, target_train, target_test = train_test_split (\
  adult_df_converted_cols.drop ('income-over-50k', axis = 1, inplace = False), adult_df_converted_cols['income-over-50k'], random_state = 16,\
)

In machine learning, the Receiver Operating Characteristic (ROC) curve and the Area Under the Curve (AUC) metric are commonly used to evaluate the performance of binary classification models. These are valuable tools for understanding how well a model can distinguish between the positive and negative classes in a classification problem.

Let's break down each concept:

1. ROC Curve:
The ROC curve is a graphical representation of a binary classifier's performance as the discrimination threshold is varied. It is created by plotting the True Positive Rate (TPR) against the False Positive Rate (FPR) at different threshold values. Here's how these rates are calculated:

- True Positive Rate (TPR), also called Sensitivity or Recall, measures the proportion of actual positive instances that are correctly identified by the model:
   TPR = True Positives / (True Positives + False Negatives)

- False Positive Rate (FPR) measures the proportion of actual negative instances that are incorrectly classified as positive by the model:
   FPR = False Positives / (False Positives + True Negatives)

The ROC curve is useful because it shows how well a classifier can trade off between sensitivity and specificity across different threshold values. The ideal ROC curve hugs the top-left corner, indicating a high TPR and low FPR, meaning the classifier is excellent at distinguishing between the two classes.

2. AUC (Area Under the Curve):
The AUC is a scalar metric derived from the ROC curve. It represents the area under the ROC curve, hence its name. The AUC ranges from 0 to 1, where 0 indicates a very poor classifier (always predicting the opposite class) and 1 signifies a perfect classifier (making all correct predictions).

The AUC metric is beneficial because it provides a single value to summarize the classifier's overall performance across all possible threshold values. It is particularly useful when dealing with imbalanced datasets, where one class significantly outnumbers the other. In such cases, accuracy alone might not be a reliable evaluation metric, and AUC can provide a more robust performance measure.

A quick rule of thumb for interpreting AUC values:
- AUC ≈ 0.5: The model performs no better than random guessing.
- 0.5 < AUC < 0.7: The model has poor to fair performance.
- 0.7 < AUC < 0.9: The model has good to excellent performance.
- AUC ≈ 1: The model is close to or has a perfect performance.

Here are your tasks:

  1. Use Scikit-Learn's [roc_auc_score](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.roc_auc_score.html) to calculate the AUC score for a method that always predicts the majority class.  

In [105]:
# Use Scikit-Learn's roc_auc_score to calculate the AUC score for a method that
# always predicts the majority class.
def always_predicts_majority_baseline (target_col):
  frequencies_list = target_col.value_counts().to_list()
  return frequencies_list.index (max (frequencies_list))

baseline_model = always_predicts_majority_baseline (target_train)
print (f"The baseline model will always predict {baseline_model} for adult income. >= $50k.")
roc_auc_score (target_test, [baseline_model] * len (target_test))

The baseline model will always predict 0 for adult income. >= $50k.


np.float64(0.5)

The area under the curve for our baseline model is 0.5, which means it exactly as useful as random guessing.

Now, let's do a little feature engineering.

1. Use Scikit-Learn's [ColumnTransformer](https://scikit-learn.org/stable/modules/generated/sklearn.compose.ColumnTransformer.html) to apply One Hot Encoding to the categorical variables in `workclass`, `education`, `marital-status`, `occupation`, `relationship`, 'race', `sex`, and `native-country`. Also, apply [MinMaxScaler](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html) to the remaining continuous features. How many columns will the dataframe have after these columns transformations are applied?

In [106]:
# Use Scikit-Learn's ColumnTransformer to apply One Hot Encoding to the
# categorical variables in workclass, education, marital-status, occupation,
# relationship, 'race', sex, and native-country.
# KRISTIAN_NOTE - Actually, let's NOT apply the OneHotEncoder to any columns.
# Because we already encoded the columns via get_dummies prior to splitting the dataset.
# We will still apply MinMaxScaler to the remaining continuous features.

all_columns_preparer = ColumnTransformer ([\
  # KRISTIAN_NOTE - For this one-hot encoder, I will set handle_unknown to 'error'
  # because there should not be any unknown data.  We haven't split the data yet.
  ('one_hot_categorical_columns', OneHotEncoder (handle_unknown = 'error'), []), #categorical_cols),
  ('scale_quantifiable_columns', MinMaxScaler (), quanitative_feature_columns)
])

In [111]:
# How many columns will the dataframe have after these columns transformations are applied?
# KRISTIAN_NOTE - We already dropped the target 'income-over-50k' column via the label-encoding above.
# Also, keep in mind the result is NOT a dataframe, but a sparse matrix.
encoded_adult_sparse_matrix = all_columns_preparer.fit_transform (features_train, target_train)
print (features_train.shape) # KRISTIAN_TODO - This has 104 columns, so why does the column transformer remove so many?
print (type (encoded_adult_sparse_matrix))
print (f"Number of rows in one-hot encoded adult dataset: {encoded_adult_sparse_matrix.shape[0]}")
print (f"Number of columns in one-hot encoded adult dataset: {encoded_adult_sparse_matrix.shape[1]}")

(22621, 104)
<class 'numpy.ndarray'>
Number of rows in one-hot encoded adult dataset: 22621
Number of columns in one-hot encoded adult dataset: 6


Same number of rows as before, so that checks out.  Now we still have 105 columns.

KRISTIAN_NOTE - The output of the ColumnTransformer is NOT a dataframe, but a CSR (Compressed Sparse Row) Matrix: https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csr_matrix.html

In [20]:
print (encoded_adult_sparse_matrix)

<Compressed Sparse Row sparse matrix of dtype 'float64'
	with 373036 stored elements and shape (30162, 105)>
  Coords	Values
  (0, 5)	1.0
  (0, 16)	1.0
  (0, 27)	1.0
  (0, 30)	1.0
  (0, 45)	1.0
  (0, 54)	1.0
  (0, 56)	1.0
  (0, 95)	1.0
  (0, 98)	0.30136986301369856
  (0, 99)	0.04333771149798496
  (0, 100)	0.8
  (0, 101)	0.021740217402174022
  (0, 103)	0.39795918367346933
  (1, 4)	1.0
  (1, 16)	1.0
  (1, 25)	1.0
  (1, 33)	1.0
  (1, 44)	1.0
  (1, 54)	1.0
  (1, 56)	1.0
  (1, 95)	1.0
  (1, 98)	0.4520547945205479
  (1, 99)	0.047277379845214206
  (1, 100)	0.8
  (1, 103)	0.12244897959183672
  :	:
  (30160, 18)	1.0
  (30160, 27)	1.0
  (30160, 30)	1.0
  (30160, 47)	1.0
  (30160, 54)	1.0
  (30160, 56)	1.0
  (30160, 95)	1.0
  (30160, 98)	0.0684931506849315
  (30160, 99)	0.12762010039865773
  (30160, 100)	0.5333333333333333
  (30160, 103)	0.19387755102040813
  (30161, 3)	1.0
  (30161, 18)	1.0
  (30161, 25)	1.0
  (30161, 33)	1.0
  (30161, 49)	1.0
  (30161, 54)	1.0
  (30161, 55)	1.0
  (30161, 95)	1.

Keras is an open-source deep learning library written in Python. It was developed to provide a user-friendly, high-level interface for building and training neural networks. The library was created by François Chollet and was first released in March 2015 as part of the Deeplearning4j project. Later, it became part of the TensorFlow ecosystem and is now the official high-level API for TensorFlow.

Keras is designed to be modular, user-friendly, and easy to extend. It allows researchers and developers to quickly prototype and experiment with various deep learning models. One of the primary goals of Keras is to enable fast experimentation, making it simple to build and iterate on different architectures.

Key features of Keras include:

1. User-friendly API: Keras provides a simple and intuitive interface for defining and training deep learning models. Its design philosophy focuses on ease of use and clarity of code.

2. Modularity: Models in Keras are built as a sequence of layers, and users can easily stack, merge, or create complex architectures using a wide range of predefined layers.

3. Extensibility: Keras allows users to define custom layers, loss functions, and metrics. This flexibility enables researchers to experiment with new ideas and algorithms seamlessly.

4. Backends: Initially, Keras supported multiple backends, including TensorFlow, Theano, and CNTK. However, as of TensorFlow version 2.0, TensorFlow has become the primary backend for Keras.

5. Multi-GPU and distributed training: Keras supports training models on multiple GPUs and in distributed computing environments, making it suitable for large-scale experiments.

6. Pre-trained models: Keras includes a collection of pre-trained models for common tasks, such as image classification (e.g., VGG, ResNet, MobileNet) and natural language processing (e.g., Word2Vec, GloVe).

The integration of Keras into TensorFlow as its official high-level API has solidified its position as one of the most popular deep learning libraries in the machine learning community. Its ease of use and versatility have contributed to its widespread adoption in both academia and industry for a wide range of deep learning tasks.

Here are your tasks:

1. Create your own model in Keras to predict income in the Adult training data. Remember, it's always better to start simple and add complexity to the model if necessary. What's a good loss function to use?

2. Keras can be integrated with Scitkit-Learn using a wrapper. Use the [KerasClassifier wrapper](https://adriangb.com/scikeras/stable/generated/scikeras.wrappers.KerasClassifier.html) to integrate your Keras model with the ColumnTransformer from previous steps using a [Pipeline](https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html) object.

3. Fit your model.

4. Calculate the AUC score of your model on the test data. Does the model predict better than random?

5. Generate an ROC curve for your model using [RocCurveDisplay](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.RocCurveDisplay.html). What would the curve look like if all your predictions were randomly generated? What would the curve look like if it you had a perfect model?

In [ ]:
# Define the Keras model

In [ ]:
# Create a Keras classifier

In [ ]:
# Create the scikit-learn pipeline

In [ ]:
# Fit the pipeline on the training data

In [ ]:
# Calculate the AUC score of your model on the test data.
# Does the model predict better than random?

In [ ]:
# Generate an ROC curve for your model.